pkg/client-sdk/covenantless_client.go

```diff
@@ -329,6 +334,155 @@ func (a *covenantlessArkClient) listenForArkTxs(ctx context.Context) {
 	}
 }
 
+func (a *covenantlessArkClient) refreshTxDb(ctx context.Context) error {
+	spendableVtxos, spentVtxos, err := a.ListVtxos(ctx)
+	if err != nil {
+		return err
+	}
+
+	boardingTxs, roundsToIgnore, err := a.getBoardingTxs(ctx)
+	if err != nil {
+		return err
+	}
+
+	offchainTxs, err := vtxosToTxsCovenantless(spendableVtxos, spentVtxos, roundsToIgnore)
+	if err != nil {
+		return err
+	}
+
+	oldTxs, err := a.store.TransactionStore().GetAllTransactions(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(oldTxs) == 0 {
+		if _, err := a.store.TransactionStore().AddTransactions(ctx, offchainTxs); err != nil {
+			return err
+		}
+
+		vtxosToAdd := make([]types.Vtxo, 0, len(spendableVtxos))
+		for _, vtxo := range spendableVtxos {
+			vtxosToAdd = append(vtxosToAdd, types.Vtxo{
+				VtxoKey: types.VtxoKey{
+					Txid: vtxo.Txid,
+					VOut: vtxo.VOut,
+				},
+				PubKey:    vtxo.PubKey,
+				Amount:    vtxo.Amount,
+				RoundTxid: vtxo.RoundTxid,
+				ExpiresAt: vtxo.ExpiresAt,
+				CreatedAt: vtxo.CreatedAt,
+			})
+		}
+		if _, err := a.store.VtxoStore().AddVtxos(ctx, vtxosToAdd); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	oldTxsMap := make(map[string]types.Transaction, len(oldTxs))
+	for _, tx := range oldTxs {
+		oldTxsMap[tx.TransactionKey.String()] = tx
+	}
+
+	// add new boarding, update existing that are not confirmed
+	boardingTxsToAdd := make([]types.Transaction, 0)
+	boardingTxsToConfirm := make([]types.Transaction, 0)
+	for _, tx := range boardingTxs {
+		oldTx, exists := oldTxsMap[tx.Hex]
+		if !exists {
+			boardingTxsToAdd = append(boardingTxsToAdd, tx)
+		} else {
+			if oldTx.CreatedAt.IsZero() && !tx.CreatedAt.IsZero() {
+				boardingTxsToConfirm = append(boardingTxsToConfirm, tx)
+			}
+		}
+	}
+
+	// add new offchain, settle existing
+	txsToAdd := make([]types.Transaction, 0)
+	txsToSettle := make([]string, 0)
+	for _, tx := range offchainTxs {
+		oldTx, exists := oldTxsMap[tx.TransactionKey.String()]
+		if !exists {
+			txsToAdd = append(txsToAdd, tx)
+		} else {
+			if tx.Settled && !oldTx.Settled {
+				txsToSettle = append(txsToSettle, tx.TransactionKey.String())
+			}
+		}
+	}
+
+	allTxsToAdd := append(boardingTxsToAdd, txsToAdd...)
+	if _, err := a.store.TransactionStore().AddTransactions(ctx, allTxsToAdd); err != nil {
+		return err
+	}
+
+	for _, tx := range boardingTxsToConfirm {
+		if _, err := a.store.TransactionStore().ConfirmTransactions(
+			ctx, []string{tx.TransactionKey.String()}, tx.CreatedAt,
+		); err != nil {
+			return err
+		}
+	}
+
+	if _, err := a.store.TransactionStore().SettleTransactions(ctx, txsToSettle); err != nil {
+		return err
+	}
+
+	oldSpendableVtxos, _, err := a.store.VtxoStore().GetAllVtxos(ctx)
```

Comment:
the logic below sounds ok but i think it missed to handle a case: the vtxos fetched from db miss a SPENT vtxo fetched from the explorer. This vtxo would never be added to the db, right?

Reply:
Done